httpovergrpc/client/client.go

```diff
@@ -0,0 +1,181 @@
+}
+
+// Close closes the HTTPOverGRPC connection.
+func (t *Transport) Close() error {
+	if t.conn != nil {
+		grpclog.Infof("Closing httpovergrpc connection.")
+		return t.conn.Close()
+	}
+	return nil
+}
+
+type wrappedBody struct {
+	reader io.Reader
+	pr     *io.PipeReader
+}
+
+func (h *wrappedBody) Close() error {
+	return h.pr.Close()
+}
+
+func (h *wrappedBody) Read(p []byte) (n int, err error) {
+	return h.reader.Read(p)
+}
+
+// NewTransport returns a new Transport for making HTTP over gRPC calls.
+func NewTransport(ctx context.Context, conn *grpc.ClientConn) *Transport {
+	return &Transport{
+		ctx:    ctx,
+		conn:   conn,
+		client: pb.NewHTTPOverGRPCClient(conn),
+	}
+}
+
+// RoundTrip implements the http.RoundTripper interface.
+func (t *Transport) RoundTrip(req *http.Request) (*http.Response, error) {
+	ctx := t.ctx
+	if ctx == context.Background() || ctx == nil {
+		ctx = req.Context()
+	}
+	grpclog.Infof("%s %q", req.Method, req.URL.String())
+
+	var body []byte
+	if req.Body != nil && req.Body != http.NoBody {
+		var err error
+		if body, err = ioutil.ReadAll(req.Body); err != nil {
+			grpclog.Errorf("Error reading body: %v", err)
+			return nil, err
+		}
+	}
+
+	hReq := &pb.HTTPOverGRPCRequest{
+		Method:  req.Method,
+		Url:     req.URL.String(),
+		Headers: fromHeader(req.Header),
+		Body:    body,
+	}
+
+	pr, pw := io.Pipe()
+	go func() {
+		pw.CloseWithError(t.doRequest(ctx, pw, hReq))
+	}()
+
+	type responseOrError struct {
+		resp *http.Response
+		err  error
+	}
+	// This needs to be buffered, otherwise the goroutine will leak.
+	respCh := make(chan responseOrError, 1)
+	go func() {
+		resp, err := readResponse(pr, req)
+		respCh <- responseOrError{resp, err}
+	}()
+	timeout := func(usage string) (*http.Response, error) {
+		pr.Close()
+		grpclog.Infof("Response header timeout: %v", usage)
+		return nil, errors.New("timeout before response header received")
+	}
+	select {
+	case re := <-respCh:
+		return re.resp, re.err
+	case <-ctx.Done():
+		return timeout("ctx.Done")
+	}
+}
+
+// doRequest makes the HTTPOverGRPC HTTPRequest and handles the response.
+func (t *Transport) doRequest(ctx context.Context, w io.Writer, req *pb.HTTPOverGRPCRequest) error {
+	// gRPC deadline becomes the HTTP request timeout.
+	ctx, cancel := context.WithCancel(ctx)
+	deadline, ok := ctx.Deadline()
+	if ok {
+		grpclog.Infof("Using Transport.Timeout = %v (plus epsilon)", deadline)
+		tt := deadline.Sub(time.Now())
+		ctx, cancel = context.WithTimeout(ctx, tt+100*time.Millisecond)
+	}
+	defer cancel()
+
+	resp, err := t.client.HTTPRequest(ctx, req)
+	if err != nil {
+		return err
+	}
+
+	if _, err = w.Write(resp.Body); err != nil {
+		grpclog.Errorf("error writing response: %v", err)
+		return err
+	}
+
+	return nil
+}
+
+// readResponse reads the HTTP response from the pipe reader and returns the
+// http response.
+func readResponse(pr *io.PipeReader, req *http.Request) (resp *http.Response, err error) {
+	defer func() {
```

Comment:
There's only one place that can return an error.  Just put `pr.CloseWithError()` there?  Then you can unname the return values.